flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/tasks/RestoreStreamTaskTest.java

```diff
@@ -0,0 +1,325 @@
+import java.util.Optional;
+import java.util.concurrent.ConcurrentLinkedQueue;
+import java.util.concurrent.atomic.AtomicLong;
+
+import static org.junit.Assert.assertEquals;
+
+/**
+ * Tests ensuring correct behaviour of {@link org.apache.flink.runtime.state.ManagedInitializationContext#isRestored}
+ * method.
+ */
+public class RestoreStreamTaskTest extends TestLogger {
+	@Test
+	public void testRestore() throws Exception {
+		AcknowledgeStreamMockEnvironment environment1 = processRecords(
+			new OperatorID(42L, 42L),
+			new CounterOperator(),
+			new OperatorID(44L, 44L),
+			new CounterOperator(),
+			Optional.empty());
+
+		assertEquals(2, environment1.getCheckpointStateHandles().getSubtaskStateMappings().size());
+
+		TaskStateSnapshot stateHandles = environment1.getCheckpointStateHandles();
+
+		try {
+			AcknowledgeStreamMockEnvironment environment2 = processRecords(
+				new OperatorID(42L, 42L),
+				new RestoreCounterOperator(),
+				new OperatorID(44L, 44L),
+				new RestoreCounterOperator(),
+				Optional.of(stateHandles));
+
+			assertEquals(2, RestoreCounterOperator.RESTORE_COUNTER.get());
+		}
+		finally {
+			RestoreCounterOperator.RESTORE_COUNTER.getAndSet(0);
+		}
+	}
+
+	@Test
+	public void testRestoreWithNewId() throws Exception {
+		AcknowledgeStreamMockEnvironment environment1 = processRecords(
+			new OperatorID(42L, 42L),
+			new CounterOperator(),
+			new OperatorID(44L, 44L),
+			new CounterOperator(),
+			Optional.empty());
+
+		assertEquals(2, environment1.getCheckpointStateHandles().getSubtaskStateMappings().size());
+
+		TaskStateSnapshot stateHandles = environment1.getCheckpointStateHandles();
+
+		try {
+			AcknowledgeStreamMockEnvironment environment2 = processRecords(
+				new OperatorID(4242L, 4242L),
+				new RestoreCounterOperator(),
+				new OperatorID(44L, 44L),
+				new RestoreCounterOperator(),
+				Optional.of(stateHandles));
+
+			assertEquals(1, RestoreCounterOperator.RESTORE_COUNTER.get());
+		}
+		finally {
+			RestoreCounterOperator.RESTORE_COUNTER.getAndSet(0);
+		}
+	}
+
+	@Test
+	public void testRestoreAfterScaleUp() throws Exception {
+		OperatorID headOperatorID = new OperatorID(42L, 42L);
+		OperatorID tailOperatorID = new OperatorID(44L, 44L);
+
+		AcknowledgeStreamMockEnvironment environment1 = processRecords(
+			headOperatorID,
+			new CounterOperator(),
+			tailOperatorID,
+			new CounterOperator(),
+			Optional.empty());
+
+		assertEquals(2, environment1.getCheckpointStateHandles().getSubtaskStateMappings().size());
+
+		// test empty state in case of scale up
+		OperatorSubtaskState emptyHeadOperatorState = StateAssignmentOperation.operatorSubtaskStateFrom(
+			new OperatorInstanceID(0, headOperatorID),
+			Collections.emptyMap(),
+			Collections.emptyMap(),
+			Collections.emptyMap(),
+			Collections.emptyMap());
+
+		TaskStateSnapshot stateHandles = environment1.getCheckpointStateHandles();
+		stateHandles.putSubtaskStateByOperatorID(headOperatorID, emptyHeadOperatorState);
```

Comment:
One point that this test will not catch is when the behaviour of the state assignment changes or something else is used instead, e.g. producing a `null` instead of an empty `OperatorSubtaskState` and this will influence the result. I am aware that covering this probably means having an IT case or covering this with another unit test - so do you think it is worth having to also guard against this kind of code modifications?

Reply:
The same also applies to a certain degree to all other tests: right now we assume that the reported states get passed back without any modification from the checkpoint coordinator. Again, modifications to that could change behaviour without detection from this test.

Reply:
ad 1. - that's why I have used `StateAssignmentOperation.operatorSubtaskStateFrom`, to share this logic. I didn't want to use all of the `StateAssignmentOperation`, because it's constructor is really annoying to fulfil.

ad 2. - that's a valid concern, however writing ITCases for this might be also an overkill. And wouldn't it be to necessary to test it against every state backend, to make sure that there are no introduced quirks during (de)serialisation?

Reply:
I would say it's great to have this unit test, eventually an IT case could also make sense. But this can be done in a different work / PR.